File: Leetcode/CountUnguardedCellsintheGrid.rs
// https://leetcode.com/problems/count-unguarded-cells-in-the-grid

use std::collections::BTreeMap;

struct Solution {}

impl Solution {
    pub fn count_unguarded(m: i32, n: i32, guards: Vec<Vec<i32>>, walls: Vec<Vec<i32>>) -> i32 {
        let (n, m) = (m as usize, n as usize);

        let mut rows = vec![BTreeMap::<usize, usize>::new(); n];
        let mut cols = vec![BTreeMap::<usize, usize>::new(); m];

        for guard in guards.iter() {
            let (x, y) = (guard[0] as usize, guard[1] as usize);
            rows[x].insert(y, 1);
            cols[y].insert(x, 1);
        }
        for wall in walls.into_iter() {
            let (x, y) = (wall[0] as usize, wall[1] as usize);
            rows[x].insert(y, 2);
            cols[y].insert(x, 2);
        }
        for i in 0..n {
            for j in 0..m {
                if !rows[i].contains_key(&j) {
                    rows[i].insert(j, 0);
                    cols[j].insert(i, 0);
                }
            }
        }

        for guard in guards.into_iter() {
            let (x, y) = (guard[0] as usize, guard[1] as usize);
            {
                let mut to_remove = vec![];
                let mut it = rows[x].range((y + 1)..);
                while let Some((&key, &value)) = it.next() {
                    if value == 0 {
                        to_remove.push(key);
                        cols[key].remove(&x);
                    } else {
                        break;
                    }
                }

                let mut it = rows[x].range(..y);
                while let Some((&key, &value)) = it.next_back() {
                    if value == 0 {
                        to_remove.push(key);
                        cols[key].remove(&x);
                    } else {
                        break;
                    }
                }

                for remove in to_remove {
                    rows[x].remove(&remove);
                }
            }
            {
                let mut to_remove = vec![];
                let mut it = cols[y].range((x + 1)..);
                while let Some((&key, &value)) = it.next() {
                    if value == 0 {
                        to_remove.push(key);
                        rows[key].remove(&y);
                    } else {
                        break;
                    }
                }

                let mut it = cols[y].range(..x);
                while let Some((&key, &value)) = it.next_back() {
                    if value == 0 {
                        to_remove.push(key);
                        rows[key].remove(&y);
                    } else {
                        break;
                    }
                }

                for remove in to_remove {
                    cols[y].remove(&remove);
                }
            }
        }

        let mut ans = 0;
        for i in 0..n {
            for (k, t) in rows[i].iter() {
                if t == &0 {
                    ans += 1;
                }
            }
        }

        ans
    }
}

File: Leetcode/CountUnguardedCellsintheGrid.cpp
// https://leetcode.com/problems/count-unguarded-cells-in-the-grid

class Solution {
   public:
    int countUnguarded(int n,
                       int m,
                       vector<vector<int>>& guards,
                       vector<vector<int>>& walls) {
        using cell = pair<int, int>;

        enum CellType { Guard = 0, Wall = 1, Fake = 42 };

        constexpr auto ma = numeric_limits<int>::max();
        constexpr auto mi = numeric_limits<int>::min();

        vector<set<cell>> rows(n);
        vector<set<cell>> cols(m);
        for (auto i = 0; i < n; ++i) {
            rows[i].insert({mi, CellType::Fake});
            rows[i].insert({ma, CellType::Fake});
        }
        for (auto j = 0; j < m; ++j) {
            cols[j].insert({mi, CellType::Fake});
            cols[j].insert({ma, CellType::Fake});
        }

        for (const auto& guard : guards) {
            const auto row = guard[0];
            const auto col = guard[1];

            rows[row].insert({col, CellType::Guard});
            cols[col].insert({row, CellType::Guard});
        }
        for (const auto& wall : walls) {
            const auto row = wall[0];
            const auto col = wall[1];

            rows[row].insert({col, CellType::Wall});
            cols[col].insert({row, CellType::Wall});
        }

        auto ans = 0;
        for (auto i = 0; i < n; ++i) {
            for (auto j = 0; j < m; ++j) {
                auto col = rows[i].lower_bound({j, mi});
                if (col->first == j) {
                    continue;
                }

                auto row = cols[j].lower_bound({i, mi});
                if (col->second != CellType::Guard &&
                    (--col)->second != CellType::Guard &&
                    row->second != CellType::Guard &&
                    (--row)->second != CellType::Guard) {
                    ++ans;
                }
            }
        }

        return ans;
    }
};